File: frontend/src/lib/guides.ts
import "server-only";

import { supabase } from "@/lib/supabase";

export type GuideVideo = {
  title: string | null;
  url: string;
  channel: string | null;
};

export type GuideImage = {
  id: string | null;
  type: string | null;
  description: string | null;
  url: string | null;
};

export type GuideSummary = {
  slug: string;
  title: string;
  summary: string;
  cover_url: string | null;
  category: string;
  difficulty: string;
  tags: string[];
  published_at: string | null;
};

export type GuideDetail = GuideSummary & {
  body: string;
  tldr: string[];
  videos: GuideVideo[];
  images: GuideImage[];
  evaluation_score: number | null;
  author_name: string | null;
};

const LIST_COLUMNS =
  "slug,title,summary,cover_url,category,difficulty,tags,published_at";
const DETAIL_COLUMNS = `${LIST_COLUMNS},body,tldr,videos,images,evaluation_score,author_name`;

function asArray<T>(value: unknown): T[] {
  return Array.isArray(value) ? (value as T[]) : [];
}

export async function fetchGuides(): Promise<GuideSummary[]> {
  const { data, error } = await supabase
    .from("guides")
    .select(LIST_COLUMNS)
    .eq("status", "published")
    .order("published_at", { ascending: false });
  if (error) throw new Error(`Failed to fetch guides: ${error.message}`);

  return (data ?? []).map((row) => ({
    slug: row.slug as string,
    title: (row.title as string) ?? "",
    summary: (row.summary as string) ?? "",
    cover_url: (row.cover_url as string | null) ?? null,
    category: (row.category as string) ?? "",
    difficulty: (row.difficulty as string) ?? "",
    tags: asArray<string>(row.tags),
    published_at: (row.published_at as string | null) ?? null,
  }));
}

export async function fetchGuide(slug: string): Promise<GuideDetail | null> {
  const safe = decodeURIComponent(slug);
  const { data, error } = await supabase
    .from("guides")
    .select(DETAIL_COLUMNS)
    .eq("slug", safe)
    .eq("status", "published")
    .limit(1)
    .maybeSingle();
  if (error) {
    if (error.code === "PGRST116") return null;
    throw new Error(`Failed to fetch guide ${slug}: ${error.message}`);
  }
  if (!data) return null;

  return {
    slug: data.slug as string,
    title: (data.title as string) ?? "",
    summary: (data.summary as string) ?? "",
    cover_url: (data.cover_url as string | null) ?? null,
    category: (data.category as string) ?? "",
    difficulty: (data.difficulty as string) ?? "",
    tags: asArray<string>(data.tags),
    published_at: (data.published_at as string | null) ?? null,
    body: (data.body as string) ?? "",
    tldr: asArray<string>(data.tldr),
    videos: asArray<GuideVideo>(data.videos),
    images: asArray<GuideImage>(data.images),
    evaluation_score: (data.evaluation_score as number | null) ?? null,
    author_name: (data.author_name as string | null) ?? null,
  };
}

File: backend/app/api/v1/guides.py
from datetime import datetime
from typing import Any

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from app.core.supabase import get_supabase

router = APIRouter(prefix="/guides", tags=["guides"])


class GuideVideo(BaseModel):
    title: str | None = None
    url: str
    channel: str | None = None


class GuideImage(BaseModel):
    id: str | None = None
    type: str | None = None
    description: str | None = None
    url: str | None = None


class GuideSummary(BaseModel):
    slug: str
    title: str
    summary: str
    cover_url: str | None = None
    category: str = ""
    difficulty: str = ""
    tags: list[str] = []
    published_at: datetime | None = None


class GuideDetail(GuideSummary):
    body: str = ""
    tldr: list[str] = []
    videos: list[GuideVideo] = []
    images: list[GuideImage] = []
    evaluation_score: float | None = None
    author_name: str | None = None


_LIST_COLUMNS = "slug,title,summary,cover_url,category,difficulty,tags,published_at"
_DETAIL_COLUMNS = (
    _LIST_COLUMNS
    + ",body,tldr,videos,images,evaluation_score,author_name"
)


def _coerce_list(value: Any) -> list:
    return value if isinstance(value, list) else []


@router.get("", response_model=list[GuideSummary])
async def list_guides() -> list[GuideSummary]:
    response = (
        get_supabase()
        .table("guides")
        .select(_LIST_COLUMNS)
        .eq("status", "published")
        .order("published_at", desc=True)
        .execute()
    )
    out: list[GuideSummary] = []
    for row in response.data or []:
        out.append(GuideSummary(
            slug=row["slug"],
            title=row.get("title", ""),
            summary=row.get("summary", ""),
            cover_url=row.get("cover_url"),
            category=row.get("category", ""),
            difficulty=row.get("difficulty", ""),
            tags=_coerce_list(row.get("tags")),
            published_at=row.get("published_at"),
        ))
    return out


@router.get("/{slug}", response_model=GuideDetail)
async def get_guide(slug: str) -> GuideDetail:
    response = (
        get_supabase()
        .table("guides")
        .select(_DETAIL_COLUMNS)
        .eq("slug", slug)
        .eq("status", "published")
        .limit(1)
        .execute()
    )
    rows = response.data or []
    if not rows:
        raise HTTPException(status_code=404, detail="guide not found")
    r = rows[0]
    return GuideDetail(
        slug=r["slug"],
        title=r.get("title", ""),
        summary=r.get("summary", ""),
        cover_url=r.get("cover_url"),
        category=r.get("category", ""),
        difficulty=r.get("difficulty", ""),
        tags=_coerce_list(r.get("tags")),
        published_at=r.get("published_at"),
        body=r.get("body", ""),
        tldr=_coerce_list(r.get("tldr")),
        videos=_coerce_list(r.get("videos")),
        images=_coerce_list(r.get("images")),
        evaluation_score=r.get("evaluation_score"),
        author_name=r.get("author_name"),
    )
